### PR TITLE
Instrument metrics in LeaseBasedSourceCoordinator

### PR DESCRIPTION
**Pull request by graytaylor0**
### Description
Instruments a number of metrics to give insight into progress tracking and failures in the source coordinator
 
### Issues Resolved
Related to #2412 
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by kkondaka — APPROVED** [review]
Looks good.

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/sourcecoordination/LeaseBasedSourceCoordinator.java [hunk 1]
```diff
@@ -170,8 +226,23 @@ public void closePartition(final String partitionKey, final Duration reopenAfter
             itemToUpdate.setReOpenAt(Instant.now().plus(reopenAfter));
         }
 
+        try {
+            sourceCoordinationStore.tryUpdateSourcePartitionItem(itemToUpdate);
+        } catch (final PartitionUpdateException e) {
+            if (SourcePartitionStatus.COMPLETED.equals(itemToUpdate.getSourcePartitionStatus())) {
+                completePartitionUpdateErrorCounter.increment();
+            } else {
+                closePartitionUpdateErrorCounter.increment();
```
Comment (cmanning09) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/sourcecoordination/LeaseBasedSourceCoordinator.java:235]: This logic won't extend well if we add a new status that is supported in this method. SaveProgressState does not have a metric for each partition status type. Do we need to differentiate between Closed and Completed? I would recommend creating a dimension that keyed off of status if we need to.

Comment (graytaylor0) [reply data-prepper-core/src/main/java/org/opensearch/dataprepper/sourcecoordination/LeaseBasedSourceCoordinator.java:235]: The only updates are saveState, complete, and close. I chose to give granularity on which action failed to update by separating to three different metrics. I am fine with adding a single metric for `partitionUpdateErrors` with a dimension that shows the action, but how does this work with micrometer

**Review by cmanning09 — APPROVED** [review]
Don't worry about changing if you don't think its worth it. I am just advocating for code that can be easily extended beyond what we can plan and know of as possibilities right now. It's a simple adjustment that will make it easier for future engineers to inherently get metrics but its not required.

**Pull request merged**